### PR TITLE
Update Tokio to use `std::future`.

### PR DESCRIPTION
**Pull request by carllerche**
Remove usage of 0.1 futures and switch to `std::future`.

Work in progress. Additional changes will be submitted via PRs to this branch.

### CODE REVIEW

File: tokio-io/src/async_write.rs [hunk 1]
```diff
@@ -143,12 +143,25 @@ pub trait AsyncWrite {
     }
 }
 
-/*
-impl<T: ?Sized + AsyncWrite> AsyncWrite for Box<T> {
-    fn shutdown(&mut self) -> Poll<(), std_io::Error> {
-        (**self).shutdown()
+impl<T: ?Sized + Unpin + AsyncWrite> AsyncWrite for Box<T> {
```
Comment (qm3ster) [inline tokio-io/src/async_write.rs]: crazy nit, but maybe the `AsyncX` and `Unpin` should be sorted in a consistent order?

**Comment by quininer** [comment]
I noticed that `tokio-io` and `tokio-futures` have some features similar to `futures-io` and `futures-sink`. Will this be a long-term split?

**Comment by carllerche** [comment]
@quininer I have not been following developments in the futures crate nor am I aware of any of their intent w.r.t. Tokio. As of now, I intend to keep things in Tokio. Once things settle, we can evaluate.

**Comment by quininer** [comment]
I think we should establish contact with `futures-rs` as early as possible. @cramertj what do you think?

**Comment by qm3ster** [comment]
Perhaps it's best to first convert to what the [`futures` compatibility layer](https://rust-lang-nursery.github.io/futures-rs/blog/2019/04/18/compatibility-layer.html) would produce for the tokio constructs, release that, and only then consider what breaking changes to make to simplify api (such as removing some errors).

They must be suitable, otherwise these conversions for io and streams wouldn't have been possible.

**Comment by cramertj** [comment]
@quininer Yes, `futures-io` covers the definition of the `AsyncRead`/`AsyncWrite`/`AsyncBufRead`/`AsyncSeek` traits, and `futures-util` provides the combinators for them. i'd expect that the corresponding parts of `tokio-io` could be replaced with these.

With respect to `tokio-futures`, it looks like it's a re-definition of the `ready!` macro, the `Sink` trait (from `futures-sink`, and a re-export of `core::future::Future` and `futures_core::stream::Stream`. I don't have any particular opinion on this, although it would be nice to use the same `Sink` trait.

**Comment by taiki-e** [comment]
@cramertj 
> it looks like it's a re-definition of the `ready!` macro,

I think we can move `ready!` macro to `futures-core` from `futures-util` (It is stable, imo).

**Comment by cramertj** [comment]
@taiki-e Sounds good to me!

File: tokio-tcp/src/stream.rs [hunk 2]
```diff
@@ -721,68 +711,17 @@ impl TcpStream {
...
-    fn shutdown(&mut self) -> Poll<(), io::Error> {
-        <&TcpStream>::shutdown(&mut &*self)
-    }
-
-    fn write_buf<B: Buf>(&mut self, buf: &mut B) -> Poll<usize, io::Error> {
-        <&TcpStream>::write_buf(&mut &*self, buf)
-    }
-}
-
-// ===== impl Read / Write for &'a =====
```
Comment (jethrogb) [inline tokio-tcp/src/stream.rs]: Why were these removed?
